core/jvmMain/src/kotlinx/serialization/internal/Caching.kt

```diff
@@ -0,0 +1,124 @@
+    return if (useClassValue) ClassValueCache(factory) else ConcurrentHashMapCache(factory)
+}
+
+/**
+ * Creates a **strongly referenced** cache of values associated with [Class].
+ * Serializers are computed using provided [factory] function.
+ *
+ * `null` values are not supported, though there aren't any technical limitations.
+ */
+internal actual fun createParametrizedCache(factory: (KClass<Any>, List<KType>) -> KSerializer<Any>?): ParametrizedSerializerCache {
+    return if (useClassValue) ClassValueParametrizedCache(factory) else ConcurrentHashMapParametrizedCache(factory)
+}
+
+@SuppressAnimalSniffer
+private class ClassValueCache(private val compute: (KClass<*>) -> KSerializer<Any>?) : SerializerCache {
+    private val classValue = initClassValue()
+
+    private fun initClassValue() = object : ClassValue<CacheEntry>() {
+        override fun computeValue(type: Class<*>): CacheEntry {
+            val pair = SerializerPair.from(compute(type.kotlin))
+            return CacheEntry(pair)
+        }
+    }
+
+    override fun get(key: KClass<Any>, isNullable: Boolean): KSerializer<Any?>? =
+        classValue[key.java].serializers?.get(isNullable)
+}
+
+@SuppressAnimalSniffer
+private class ClassValueParametrizedCache(private val compute: (KClass<Any>, List<KType>) -> KSerializer<Any>?) : ParametrizedSerializerCache {
+    private val classValue = initClassValue()
+
+    private fun initClassValue() = object : ClassValue<ParametrizedCacheEntry>() {
+        override fun computeValue(type: Class<*>): ParametrizedCacheEntry {
+            return ParametrizedCacheEntry()
+        }
+    }
+
+    override fun get(key: KClass<Any>, isNullable: Boolean, types: List<KType>): Result<KSerializer<Any?>?> =
+        classValue[key.java]
+            .computeIfAbsent(types) { compute(key, types) }
+            .map { it?.get(isNullable) }
+}
+
+/**
+ * We no longer support Java 6, so the only place we use this cache is Android, where there
+ * are no classloader leaks issue, thus we can safely use strong references and do not bother
+ * with WeakReference wrapping.
+ */
+private class ConcurrentHashMapCache(private val compute: (KClass<*>) -> KSerializer<Any>?) : SerializerCache {
+    private val cache = ConcurrentHashMap<Class<*>, CacheEntry>()
+
+    override fun get(key: KClass<Any>, isNullable: Boolean): KSerializer<Any?>? {
+        return cache.getOrPut(key.java) {
+            CacheEntry(SerializerPair.from(compute(key)))
+        }.serializers?.get(isNullable)
+    }
+}
+
+
+
+private class ConcurrentHashMapParametrizedCache(private val compute: (KClass<Any>, List<KType>) -> KSerializer<Any>?) : ParametrizedSerializerCache {
+    private val cache = ConcurrentHashMap<Class<*>, ParametrizedCacheEntry>()
+
+    override fun get(key: KClass<Any>, isNullable: Boolean, types: List<KType>): Result<KSerializer<Any?>?> {
+        return cache.getOrPut(key.java) { ParametrizedCacheEntry() }
```

Comment:
Two `.getOrPut` in a row doesn't look very fast. Maybe we need additional benchmarks (with 2,3,4 or more type params) to determine if we need the parameterized cache at all.

Again, I think most of the cases for top-level generic serializers would be `List<Something>()` or `MyGeneric<X>()` where there are at most several (up to 10) different X so probably specialized solution may be more efficient.

Reply:
Regardless on decision on this particular feature, we also probably need some kind of benchmarks for memory footprint/throughput

Reply:
I added tests for generic class `class PentaGeneric<T1, T2, T3, T4, T5>(val a: T1, val b: T2, val c: T3, val d: T4, val e: T5)` and got the following results:
with cache
```
LookupOverheadBenchmark.pentaGenericPlain     thrpt   14  3.285 ± 0.077  ops/us
LookupOverheadBenchmark.pentaGenericReified   thrpt   14  2.763 ± 0.042  ops/us
```

without cache
```
LookupOverheadBenchmark.pentaGenericPlain     thrpt   14  3.318 ± 0.017  ops/us
LookupOverheadBenchmark.pentaGenericReified   thrpt   14  0.894 ± 0.007  ops/us
```

So I think it makes sense to use a cache even with a large number of types